completion_aggregator/models.py

```diff
@@ -1,27 +1,189 @@
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * aggregation_name (string): The name of the aggregated blocks.
+              This is set by the level that the aggregation
+              is occurring. Possible values are "course", "chapter", "sequential", "vertical"
+            * earned (float): The positive sum of the fractional completions of all descendant blocks up
+              to the value of possible.
+            * possible (float): The total sum of the possible completion values of all descendant
+              blocks that are visible to the user. This should be a positive integer.
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises:
+            TypeError:
+                If the wrong type is passed for the parameters.
+
+            ValueError:
+                If the value of earned is greater than possible.
+
+            django.core.exceptions.ValidationError:
+                If earned / possible results in a number that is less than 0 or greater than 1 or any float is
+                less than zero.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+        self.validate(user, course_key, block_key)
+        if earned > possible:
+            raise ValueError(_('Earned cannot be greater than the possible value.'))
+        if possible > 0.0:
+            percent = earned / possible
+        else:
+            percent = 1.0
+        obj, is_new = self.update_or_create(
+            user=user,
+            course_key=course_key,
+            aggregation_name=aggregation_name,
+            block_key=block_key,
+            defaults={
+                'percent': percent,
+                'possible': possible,
+                'earned': earned,
+            },
+        )
+        return obj, is_new
+
 
 @python_2_unicode_compatible
-class AggregateCompletion(TimeStampedModel):
+class Aggregator(TimeStampedModel):
     """
-    TODO: replace with a brief description of the model.
+    Aggregators are blocks that contain other blocks, are not themselves completable,
+    and are considered 100% complete when all descendant blocks are complete.
     """
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    aggregation_name = models.CharField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
```

Comment:
Please remove `max_length` here , since it's automatically added by the `OpaqueKeyField` class.